### PR TITLE
Add compilable and capturable foreach adamax with tests

### LINKED ISSUE #117812: Support full graph compilation of foreach Adamax
### 🚀 The feature, motivation and pitch

Refactor Adamax to support multitensor full graph compilation with torch.compile. 

WIP PR https://github.com/pytorch/pytorch/pull/110345

I'm going to take this over and land it

### Alternatives

_No response_

### Additional context

_No response_

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### PR DESCRIPTION
**Pull request by mlazos**
Based off of https://github.com/pytorch/pytorch/pull/110345

Fixes https://github.com/pytorch/pytorch/issues/117812

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @janeyx99

### CODE REVIEW

File: torch/testing/_internal/common_optimizers.py [hunk 1]
```diff
@@ -442,7 +466,7 @@ def optim_inputs_func_adamax(device=None):
             kwargs={"weight_decay": 0.9, "maximize": True},
             desc="maximize",
         ),
-    ]
+    ] + cuda_supported_configs
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:469]: feels lke there's a missed conditional for if device is cuda

File: torch/optim/adamax.py [hunk 2]
```diff
@@ -336,19 +352,35 @@ def _multi_tensor_adamax(
...
+        if capturable:
+            bias_corrections = torch._foreach_pow(beta1, grouped_state_steps)
+            # foreach_sub doesn't allow a scalar as the first arg
+            torch._foreach_sub_(bias_corrections, 1)
+
+            # foreach_div doesn't allow a scalar as the first arg
+            torch._foreach_div_(bias_corrections, lr)
+            torch._foreach_reciprocal_(bias_corrections)
+
+            numerator = torch._foreach_mul(grouped_exp_avgs, bias_corrections)
```
Comment (janeyx99) [inline torch/optim/adamax.py:381]: Does this introduce new memory? 
Could this be rewritten as the following to be explicit about memory savings?

```suggestion
            torch._foreach_mul_(bias_corrections, grouped_exp_avgs)
            numerator = bias_corrections
```

File: torch/testing/_internal/common_optimizers.py [hunk 3]
```diff
@@ -431,6 +431,30 @@ def optim_error_inputs_func_adam(device, dtype):
...
+        OptimizerInput(
+            params=None,
+            kwargs={"weight_decay": 0.9, "maximize": False, "capturable": True},
+            desc="capturable, weight_decay",
+        ),
+        OptimizerInput(
+            params=None,
+            kwargs={"weight_decay": 0, "maximize": False, "capturable": True},
+            desc="capturable",
+        ),
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:455]: this is redundant with the first line

File: torch/optim/adamax.py [hunk 4]
```diff
@@ -336,19 +352,35 @@ def _multi_tensor_adamax(
...
+            torch._foreach_addcdiv_(grouped_params, numerator, grouped_exp_infs)
+        else:
+            bias_corrections = [1 - beta1 ** _get_value(step) for step in grouped_state_steps]
+            step_size = _stack_if_compiling([(lr / bc) * -1 for bc in bias_corrections])
```
Comment (janeyx99) [inline torch/optim/adamax.py:385]: Do we still need the stack if compiling here? this path shouldn't be run by compile right?

**Review by janeyx99 — COMMENTED** [review]
Nice, left some review but overall p good!

File: torch/optim/adamax.py (hunk 2 above)
Comment (mlazos) [reply torch/optim/adamax.py:381]: yeah I tried this, but I don't think scalar tensors are supported in the first argument.

Comment (mlazos) [reply torch/optim/adamax.py:381]: Ah it's not that exact error but here it is:
https://gist.github.com/mlazos/dfa711139f0b011276622c36d2f7d74e

Comment (janeyx99) [reply torch/optim/adamax.py:381]: Ah, I see, if this passes the peak memory test I think it's fine then!

**Review by janeyx99 — APPROVED** [review]

**Review by janeyx99 — COMMENTED** [review]
Nice, could you point to/show the logs for the new compiled Adamax tests when you get a chance?

**Comment by mlazos** [comment]
> Nice, could you point to/show the logs for the new compiled Adamax tests when you get a chance?

https://gist.github.com/mlazos/4006518fa77897a9e67a6a2364e55212

**Comment by janeyx99** [comment]
Would like to callout that this change led to eager speedup with using `_foreach_maximum`:
image

The most extreme of these is a 5+x speedup for DALLE2_pytorch going from ~0.118 -> ~0.020.

**Pull request closed without merging**